src/sentry/replays/lib/query.py

```diff
@@ -114,28 +115,77 @@ def insert(self, field_name: Optional[str], value: Field) -> None:
 
 
 def generate_valid_conditions(
-    query: List[SearchFilter], query_config: QueryConfig
-) -> Generator[None, None, Condition]:
+    query: List[Union[SearchFilter, ParenExpression, str]], query_config: QueryConfig
+) -> List[Condition]:
+    """Convert search filters to snuba conditions."""
+    result = []
+    look_back = None
     for search_filter in query:
-        # Validate field exists and is filterable.
-        field_alias = search_filter.key.name
-        field = query_config.get(field_alias)
-        if field is None or not field.is_filterable:
-            raise ParseError(f"Invalid field specified: {field_alias}.")
-
-        # Validate strategy is correct.
-        query_operator = search_filter.operator
-        operator, errors = field.deserialize_operator(query_operator)
-        if errors:
-            raise ParseError(f"Invalid operator specified: {field_alias}.")
-
-        # Deserialize value to its correct type or error.
-        query_value = search_filter.value.value
-        value, errors = field.deserialize_value(query_value)
-        if errors:
-            raise ParseError(f"Invalid value specified: {field_alias}.")
-
-        yield Condition(Column(field.query_alias or field.attribute_name), operator, value)
+        # SearchFilters are appended to the result set.  If they are top level filters they are
+        # implicitly And'ed in the WHERE/HAVING clause.
+        if isinstance(search_filter, SearchFilter):
+            condition = filter_to_condition(search_filter, query_config)
+            if look_back == "AND":
+                look_back = None
+                attempt_compressed_condition(result, condition, And)
+            elif look_back == "OR":
+                look_back = None
+                attempt_compressed_condition(result, condition, Or)
+            else:
+                result.append(condition)
+        # ParenExpressions are recursively computed.  If more than one condition is returned then
+        # those conditions are And'ed.
+        elif isinstance(search_filter, ParenExpression):
+            conditions = generate_valid_conditions(search_filter.children, query_config)
+            if len(conditions) < 2:
+                result.extend(conditions)
+            else:
+                result.append(And(conditions))
+        # String types are limited to AND and OR... I think?  In the case where its not a valid
+        # look-back it is implicitly ignored.
+        elif isinstance(search_filter, str):
+            look_back = search_filter
+
+    return result
+
+
+def filter_to_condition(search_filter: SearchFilter, query_config: QueryConfig) -> Condition:
```

Comment:
wonder if it may be worth adding basic unit tests for this method / `generate_valid_conditions`

Reply:
In the future yes but right now I'm leaning towards no.  This library is only used by replays so our end-to-end coverage is sufficient.  If it were used by other teams we would want to add unit coverage to manage regressions.